File: bend2/effs/tcp_recv_text_go.js
// TCP
// ===

// See tcp_recv_text_go.c: carried bytes first, a cut scalar carried out,
// a recv of 0 bytes replaces the carry and answers need 4, and a recv that
// only feeds the carry parks again.
function tcp_recv_text_go(socket, pend, need, max, k) {
  const sys = io_sys();
  const fd = socket;
  const len = Math.min(Number(max), 2147483647);
  const b = new Uint8Array(len + 4);
  const again = sys.mac ? 35 : 11;
  let had = need < 4 ? Number(need) : 0;
  for (let m = 0; m < had; m += 1) {
    b[m] = (Number(pend) >>> (8 * m)) & 255;
  }
  const go = () => {
    const got = Number(sys.recv(fd, sys.ptr(b, had), len, 0));
    if (got < 0) {
      const code = sys.errno();
      if (code === again) {
        io_park_on(fd, false, k, go);
        return undefined;
      }
      return io_tup(socket, io_fail(code));
    }
    const eof = got === 0 && len !== 0;
    const n = had + got;
    let cut = 0, out = 0;
    for (let j = 1; !eof && j <= 3 && j <= n; j += 1) {
      const h = b[n - j];
      if ((h & 0xc0) !== 0x80) {
        const w = h >= 0xc2 && h <= 0xdf ? 2 : h >= 0xe0 && h <= 0xef ? 3
          : h >= 0xf0 && h <= 0xf4 ? 4 : 0;
        cut = w > j ? j : 0;
        break;
      }
    }
    if (got > 0 && cut === n) {
      had = n;
      io_park_on(fd, false, k, go);
      return undefined;
    }
    for (let m = 0; m < cut; m += 1) {
      out |= b[n - cut + m] << (8 * m);
    }
    return io_tup(socket, io_done(io_tup(out, eof ? 4 : cut, io_text(b, n - cut))));
  };
  return go();
}

function tcp_recv_text_go_need() {
  return { read: true };
}

File: bend2/effs/tcp_recv_text_go.c
// TCP
// ===

// TCP.recv through a decoder state, as file_read_text_go.c: the carried
// bytes (pend, low byte first; need of them) go before the ones received,
// and a scalar cut by the end of the chunk is carried out instead of
// replaced. A recv of 0 bytes is the end of the stream: the carry is
// replaced byte by byte and need answers 4. A recv that only feeds the
// carry would answer "" with the peer still open, so it parks again.
static u64 tcp_recv_text_go_cut(const char* p, u64 n) {
  for (u64 j = 1; j <= 3 && j <= n; j += 1) {
    u32 b = (u8)p[n - j];
    if ((b & 0xc0) != 0x80) {
      u64 k = b >= 0xc2 && b <= 0xdf ? 2 : b >= 0xe0 && b <= 0xef ? 3
        : b >= 0xf0 && b <= 0xf4 ? 4 : 0;
      return k > j ? j : 0;
    }
  }
  return 0;
}

static Term tcp_recv_text_go_pack(Env e, IoWork* w, u64 n, u64 cut, bool eof) {
  Term r;
  if (w->code) {
    r = io_fail(e, w->code, NULL);
  } else {
    u32 pend = 0;
    for (u64 m = 0; m < cut; m += 1) {
      pend |= (u32)(u8)w->data[n - cut + m] << (8 * m);
    }
    r = io_done(e, io_tup(e, pend, io_tup(e, eof ? 4 : (u32)cut,
      io_str(e, w->data, n - cut))));
  }
  free(w->data);
  return io_tup(e, io_hand(w->hand), r);
}

// The loop parked the request until the socket was readable; a recv that
// still finds nothing (the socket is non-blocking) parks again.
static Term tcp_recv_text_go_more(Env e, IoWork* w) {
  int fd  = (int)w->hand;
  w->size = io_sys_end(w, recv(fd, w->data + w->made, (size_t)w->word, 0));
  if (w->code == EAGAIN) {
    return io_wait_on(w, fd, POLLIN, 0, tcp_recv_text_go_more);
  }
  if (w->code) {
    return tcp_recv_text_go_pack(e, w, 0, 0, false);
  }
  bool eof = w->size == 0 && w->word != 0;
  u64  n   = (u64)w->made + w->size;
  u64  cut = eof ? 0 : tcp_recv_text_go_cut(w->data, n);
  if (w->size > 0 && cut == n) {
    w->made = (intptr_t)n;
    return io_wait_on(w, fd, POLLIN, 0, tcp_recv_text_go_more);
  }
  return tcp_recv_text_go_pack(e, w, n, cut, eof);
}

Term tcp_recv_text_go_run(Env e, Term* f, IoWork* w) {
  w->hand = (intptr_t)io_hand_v(f[0]);
  w->made = f[2] < 4 ? (intptr_t)f[2] : 0;
  w->word = f[3] < INT32_MAX ? f[3] : INT32_MAX;
  w->data = io_mem(malloc((size_t)w->word + 4));
  for (intptr_t m = 0; m < w->made; m += 1) {
    w->data[m] = (char)(f[1] >> (8 * m));
  }
  return tcp_recv_text_go_more(e, w);
}

static void __attribute__((constructor)) tcp_recv_text_go_use(void) {
  io_eff(CID_TCP_RECV_TEXT_GO, tcp_recv_text_go_run, IO_READ);
}
